femtools/Detector_Data_Types.F90

```diff
@@ -37,22 +37,43 @@ module detector_data_types
   private
   
   public :: detector_type, rk_gs_parameters, detector_linked_list, &
-            detector_list_ptr, stringlist, &
-            STATIC_DETECTOR, LAGRANGIAN_DETECTOR
+            detector_list_ptr, stringlist, attr_names_type, attr_write_type, field_phase_type, &
+            STATIC_DETECTOR, LAGRANGIAN_DETECTOR, allocate, deallocate
 
-  integer, parameter :: STATIC_DETECTOR=1, LAGRANGIAN_DETECTOR=2  
+  integer, parameter :: STATIC_DETECTOR=1, LAGRANGIAN_DETECTOR=2
 
   type stringlist
-     !!< Container type for a list of strings.
-     character(len=FIELD_NAME_LEN), dimension(:), pointer :: ptr
+    !!< Container type for a list of strings.
+    character(len=FIELD_NAME_LEN), dimension(:), pointer :: ptr
   end type stringlist
 
+  type attr_names_type
```

Comment:
I do like your suggestion, but I think I went with the more split-out types to avoid some redundancy in the objects we pass around. I guess the way I've written it, you pass a few objects like `names` and `vals` that encapsulate scalar, vector and tensor components. Your way means that we pass around the scalar, vector and tensor versions of a single descriptor.

It would be a big change to rewrite with a new data structure, so it's my mistake for picking something a bit awkward!